Add blogpost explaining `Signal` (#76)
clash-lang/clash-lang.org: content/blog/0007-signals.md

  - "Design"
---

<!-- *Cross-posted: qbaylogic.com/blog/bla* -->

I recently came across a [question on /r/haskell](https://old.reddit.com/r/haskell/comments/1hjyzfv/can_clashhaskellsupport_for_asynchronous_circuit/), where [/u/netj_nsh](https://old.reddit.com/user/netj_nsh) asked whether Clash supports asynchronous circuit designs. They went on to ask whether designing with multiple, synchronous clock domains is possible. While the (very) short answers are _no_ and _yes_, respectively, I figured I'd write a blog post clarifying these concepts and how they relate to Clash. Mostly though, I just wanted an excuse to write about Clash's simple yet clever trick that makes it tick: `Signal`.

## Combinational logic
In digital design, [combinational logic](https://en.wikipedia.org/wiki/Combinational_logic) is a type of logic circuit that has no memory. It's a circuit that computes its output based solely on its inputs. This is very similar to the concept of a pure function in functional programming: a function that always produces the same output for the same input, with no side effects. For any function in Haskell we therefore know that we can translate it to a digital design if:

> 1. The function is pure: it has no side effects. In Haskell this is enforced by the type system: the absence of `IO` in its type signature indicates it is pure<sup><a href="#footnote-1">1</a></sup>.

Reviewer: ```suggestion
1. The function is pure: it has no side effects. In Haskell this is enforced by the type system: the absence of `IO` in its type signature indicates it is pure<sup id="footnote-1-back"><a href="#footnote-1">1</a></sup>.
```

## Signals
Using the digital, sequential abstraction signals over time can be modeled as linked lists. Each element represents the stable value at each clock tick. The default list (`[]`) implementation in Haskell already gets us pretty close:

```haskell
data List a
  = []
  | a : List a
```

This allows us to write:

```haskell
>>> 1 : 2 : 3 : 4 : []
[1, 2, 3, 4]
>>> [1, 2, 3, 4] -- syntactic sugar
[1, 2, 3, 4]
```

This is very close to what we want from a signal, but we're missing two things:

* Signals don't really have an "end"
* Signals have a clock domain associated with them

Hence, if we look at the definition of Clash's `Signal` we see:

```haskell
data Signal (dom :: Domain) (a :: Type) = a :- Signal dom a
```

We can use this definition in simulation:

```haskell
>>> signal = pure 1 :: Signal System Int
>>> signal
1 :- 1 :- 1 :- 1 :- 1 :-
^C
```

To save some time, I pressed interrupt (`^C`) to prevent Haskell from printing indefinitely. Like lists, we can map over it:

```haskell
>>> fmap (+10) signal -- More typically written as `(+10) <$> signal`
11 :- 11 :- 11 :- 11 :- 11 :-
^C
```

> We can also combine multiple signals and perform some computation on them<sup><a href="#footnote-2">2</a></sup>:

Reviewer: ```suggestion
We can also combine multiple signals and perform some computation on them<sup id="footnote-2-back"><a href="#footnote-2">2</a></sup>:
```

### Signals 🤝 state
So far we've only seen combinational operations: we haven't introduced any state yet. Worse, with the introduced primitives we cannot add any state either. Let's take a step back and see what it would have to look like:

<script type="wavedrom">
{ signal: [
  { name: "clock", wave: '0p...' },
  { name: "D",    wave: "22222",  data: ["A", "B", "C", "D", "E"] },
  { name: "Q",    wave: "22222",  data: ["init", "A", "B", "C", "D", "E"] },
]}
</script>

I.e., the output signal **Q** is a shifted version of **D**. In this diagram, the initial state of the register is marked **init**. A fourth primitive Clash offers is therefore `register`:

```haskell
-- | Simplified version of Clash's `register`: the actual version has more control signals.
register :: Clock dom -> a -> Signal dom a -> Signal dom a
register _clk init signal = init :- signal
```

> Again, when Clash sees this primitive used, it generates a hardcoded bit of HDL. Note that because `register` operates on a single domain, it does not need to be aware of any properties of the clock; the abstraction works the same for clocks running on 1 MHz or 100 MHz. The generated HDL *does* need clock lines to be routed though, so the primitive still needs the clock argument.

Reviewer: ```suggestion
Again, when Clash sees this primitive being used, it generates a hardcoded bit of HDL. Note that because `register` operates on a single domain, it does not need to be aware of any properties of the clock; the abstraction works the same for clocks running on 1 MHz or 100 MHz. The generated HDL *does* need clock lines to be routed though, so the primitive still needs the clock argument.
```

And just like that, we've introduced state. We can now implement a simple counter:

```haskell
counter = register clock (fmap (+1) counter)

-- Because `Signal` has a `Num` instance, we can also write:
-- counter = register clock (counter + 1)
```

> From an HDL compilation point of view, we can see how this might work. If Clash encounters the special primitive `register`, it will not try to translate the definition, but it will insert a predefined bit of HDL.

Reviewer: ```suggestion
From an HDL compilation point of view, we can see how this might work. If Clash encounters the special primitive [`register`](https://hackage.haskell.org/package/clash-prelude-1.8.2/docs/Clash-Explicit-Signal.html#v:register), it will not try to translate the definition, but it will insert a predefined bit of HDL that implements the register.
```

> From a simulation perspective it is less clear why this works as we're using the result of a function call as its argument. In most other languages this would be either rejected at compile time or get stuck at runtime. Not so much in Haskell, as it is a [lazy language](https://en.wikipedia.org/wiki/Lazy_evaluation). In a lazy language, definitions get "unfolded" as we go. The evaluation process roughly<sup><a href="#footnote-3">3</a></sup> looks like:

Reviewer: ```suggestion
From a simulation perspective it is less clear why this works as we're using the result of a function call as its argument. In most other languages this would be either rejected at compile time or get stuck at runtime. Not so much in Haskell, as it is a [lazy language](https://en.wikipedia.org/wiki/Lazy_evaluation). In a lazy language, definitions get "unfolded" as we go. The evaluation process roughly<sup id="footnote-3-back"><a href="#footnote-3">3</a></sup> looks like:
```

  = 0 :-           let tmp = 1       :- tmp + 1 in tmp -- <=> inline tmp
  = 0 :- 1 :-      let tmp = (1 + 1) :- tmp + 1 in tmp -- <=> eval (+)
  = 0 :- 1 :-      let tmp = 2       :- tmp + 1 in tmp -- <=> inline tmp
  = 0 :- 1 :- 2 :- let tmp = (2 + 1) :- tmp + 1 in tmp -- <=> eval (+)
  = ..
```

### Signals on multiple domains
One of the questions `/u/netj_nsh` asked was: 

> Does Clash [..] support [..] synchronous design with clock domain crossing?

With the primitives introduced we cannot construct any clock domain crossing. In fact, the Haskell compiler will stop us before we even get a chance to run it:

```haskell
>>> a = pure 3 :: Signal A Int
>>> b = pure 5 :: Signal B Int
>>> liftA2 (+) a b
*** Type error: could not deduce A ~ B
```

And this is great! It means you can never accidentally cross clock domains in Clash. Of course, sometimes you *do* want to cross clock domains, which can be done with yet another primitive: [`unsafeSynchronizer`](https://hackage.haskell.org/package/clash-prelude-1.8.2/docs/Clash-Explicit-Signal.html#v:unsafeSynchronizer):

```haskell
unsafeSynchronizer :: Clock dom1 -> Clock dom2 -> Signal dom1 a -> Signal dom2 a
```

> This will reinterpret a signal on domain `dom1` as if it were a signal on `dom2`. It keeps a measure called *relative time* that keeps track of the distance to the edges. The scalar variable `relativeTime` represents the time difference between the next two edges of the two clocks: the clock of the consumed signal and the clock of the produced signal. With the scalar variable, we can distinguish two moments in time. `t = 0` points at the next edge of the consumed signal, and `t = relativeTime` points at the next edge of the produced signal. If `relativeTime` is less than zero, it means an element should now be produced. If it is greater than zero, it means an element should now be consumed. If `relativeTime` equals zero, both should happen. While the current definition is more involved, Clash used to program this as:

Reviewer: ```suggestion
This will reinterpret a signal on domain `dom1` as if it were a signal on `dom2`. It keeps a measure called *relative time* that keeps track of the distance to the edges. The scalar variable `relativeTime` represents the time difference between the next two edges of the two clocks: the clock of the consumed signal and the clock of the produced signal. With the scalar variable, we can distinguish two moments in time. `t = 0` points at the next edge of the consumed signal, and `t = relativeTime` points at the next edge of the produced signal. If `relativeTime` is less than zero, it means an element should now be produced. If it is greater than zero, it means an element should now be consumed. If `relativeTime` equals zero, both should happen. While the real definition in Clash is more involved, this computation boils down to this code:
```

    },
    {
      "name": "A => B",
      "wave": "22..2..",
      "data": [
        0,
        3,
        6
      ],
      "period": 1
    },
    {
      "name": "A => B => A",
      "wave": "2222222",
      "data": [
        0,
        3,
        3,
        3,
        6,
        6,
        6,
      ],
      "period": 1
    }
  ]
}
</script>

In the experience of yours truly, this is not a problem in practice: synchronization with `unsafeSynchronizer` is almost always backed by a memory element, sidestepping the problem entirely.

## Recap
To recap, `Signal` works for HDL translation because its implementation is hidden from the public API. To make the data type useful still, primitives are offered:

* `pure`: construct a signal
* `<$>` or `fmapSignal`: apply a pure function/combinational circuit to a signal
* `<*>` or `appSignal`: merge multiple signals on the same domain
* `register`: introduce state

For simulation it works because:

 * Lazy evaluation means we can create feedback loops without compilation or runtime errors.
 * We can "synchronize" signals on multiple domains by calculating their relative clock offsets and constructing a new signal.

## Footnotes

> <sup id="footnote-1">1</sup>This isn't entirely true: Haskell provides "escape hatches" such as `unsafePerformIO` that allow you to break purity.

Reviewer: ```suggestion
<sup id="footnote-1"><a href="#footnote-1-back">1</a></sup>This isn't entirely true: Haskell provides "escape hatches" such as `unsafePerformIO` that allow you to break purity.
```

> <!-- <sup>2</sup>In theory Clash could analyze `map`, determine how often it uses its function argument, and render that as input and outputs. At the time of writing, Clash does not do this. -->

Reviewer: Left over unused text?

> <sup id="footnote-2">2</sup>These three operations effectively implement the [Applicative](https://hackage.haskell.org/package/base-4.21.0.0/docs/Prelude.html#t:Applicative) and [Functor](https://hackage.haskell.org/package/base-4.21.0.0/docs/Prelude.html#t:Functor) typeclasses.

Reviewer: ```suggestion
<sup id="footnote-2"><a href="#footnote-2-back">2</a></sup>These three operations effectively implement the [Applicative](https://hackage.haskell.org/package/base-4.21.0.0/docs/Prelude.html#t:Applicative) and [Functor](https://hackage.haskell.org/package/base-4.21.0.0/docs/Prelude.html#t:Functor) typeclasses.
```

> <sup id="footnote-3">3</sup>The "unfolding" shown roughly follows Haskell's evaluation if the signal would be [force](https://hackage.haskell.org/package/deepseq-1.5.1.0/docs/Control-DeepSeq.html#v:force)d.

Reviewer: ```suggestion
<sup id="footnote-3"><a href="#footnote-3-back">3</a></sup>The "unfolding" shown roughly follows Haskell's evaluation if the signal would be [force](https://hackage.haskell.org/package/deepseq-1.5.1.0/docs/Control-DeepSeq.html#v:force)d.
```

2. All arguments and the return value are of a type that can be represented as a finite number of bits.
3. Any recursion, if present, should terminate at compile time. I.e., no data dependent recursion.

<!-- This doesn't mean we cannot _use_ functions that are non-representable on their own though. Take `map`:

```haskell
map ::
  (a -> b) -> -- argument 1: a function from type `a` to type `b`
  Vec n a ->  -- argument 2: a vector of `n` elements with element type `a`
  Vec n b     -- result:     a vector of `n` elements with element type `b`
```

`map` is a higher-order function that applies a function to each element of a vector. The function `a -> b` is not representable as a finite number of bits<sup>2</sup>, but we can still use `map` in our designs. For example, `map (+1)` is a function that adds one to each element of a vector. This function is representable in hardware:

```haskell
top :: Vec 3 Int -> Vec 3 Int
top vec = map (+1) vec
```

Clash will inline the definition of `map` to obtain a structure that's more obviously representable in hardware:

```haskell
top :: Vec 3 Int -> Vec 3 Int
top (e1 :> e2 :> e3 :> Nil) = (e1 + 1) :> (e2 + 1) :> (e3 + 1) :> Nil

> ``` -->

Reviewer 2: Leftover comment?;p

> Because Haskell is an excellent language to describe pure functions (subjective, I know), it should also be an excellent language to describe combinatorial logic. This leaves out a big chunk of hardware designs though: most realistic designs have lots of local state. In fact, local state is often the only way to [reach realistic clock speeds](https://en.wikipedia.org/wiki/Pipelining_(DSP_implementation)). To have any shot at becoming a good hardware description language, Clash should therefore be able to describe non-combinational (_sequential_) logic too. Still, you won't find the typical way of keeping state (e.g., the `State` monad) in a typical Clash design.

Author: Actually I do think "combinatorial" is wrong!

[Combinatorial](https://en.wikipedia.org/wiki/Combinatorics)

[Combinational](https://en.wikipedia.org/wiki/Combinational_logic)

> Because Haskell is an excellent language to describe pure functions (subjective, I know), it should also be an excellent language to describe combinatorial logic. This leaves out a big chunk of hardware designs though: most realistic designs have lots of local state. In fact, local state is often the only way to [reach realistic clock speeds](https://en.wikipedia.org/wiki/Pipelining_(DSP_implementation)). To have any shot at becoming a good hardware description language, Clash should therefore be able to describe non-combinational (_sequential_) logic too. Still, you won't find the typical way of keeping state (e.g., the `State` monad) in a typical Clash design.

Author: I'll leave out the "though" and let de colon live.

> Because Haskell is an excellent language to describe pure functions (subjective, I know), it should also be an excellent language to describe combinatorial logic. This leaves out a big chunk of hardware designs though: most realistic designs have lots of local state. In fact, local state is often the only way to [reach realistic clock speeds](https://en.wikipedia.org/wiki/Pipelining_(DSP_implementation)). To have any shot at becoming a good hardware description language, Clash should therefore be able to describe non-combinational (_sequential_) logic too. Still, you won't find the typical way of keeping state (e.g., the `State` monad) in a typical Clash design.

Reviewer 3: :+1: I think that clears it up :)

![alt text](setup-hold-1.drawio.png)

If we chain these together and mark the stable periods with a label, we get the following picture:

![alt text](setup-hold-2.drawio.png)

This is very close to how digital designers think: they think in terms of *stable periods*. Hence, an often heard phrase is: *at clock cycle **x**, the value of the signal was **y***, even though the **y** might have flipped between many values during stabilization. Condensing the previous diagram to its essentials, we get a [waveform](https://en.wikipedia.org/wiki/Waveform) diagram (in this case generated by [wavedrom](https://wavedrom.com/editor.html)):

<script type="wavedrom">
{ signal: [
  { name: "clock", wave: '0p...' },
  { name: "D",    wave: "22222",  data: ["A", "B", "C", "D", "E"] },
]}
</script>

> This abstraction is called the digital, sequential abstraction. Digital, because we don't consider any signal levels between 0 or 1. Sequential, because flip-flops can introduce memory/state.

Author: yeah very nice, thanks

## Signals
Using the digital, sequential abstraction signals over time can be modelled as linked lists. Each element represents the stable value at each clock tick. The default list (`[]`) implementation in Haskell already gets us pretty close:

```haskell
data List a
  = []
  | a : List a
```

This allows us to write:

```haskell
>>> 1 : 2 : 3 : 4 : []
[1, 2, 3, 4]

> >>> [1, 2, 3, 4] -- syntax sugar

Author: Ah bugger, I can never remember

[1, 2, 3, 4]
```

This is very close to what we want from a signal, but we're missing two things:

* Signals don't really have an "end"
* Signals have a clock domain associated with them

Hence, if we look at the definition of Clash's `Signal` we see:

```haskell
data Signal (dom :: Domain) (a :: Type) = a :- Signal dom a
```

We can use this definition in simulation:

```haskell
>>> signal = pure 1 :: Signal System Int
>>> signal
1 :- 1 :- 1 :- 1 :- 1 :-
^C
```

To save some time, I pressed interrupt (`^C`) to prevent Haskell from printing indefinitely. Like lists, we can map over it:

```haskell

> >>> fmap (+10) signal -- More typically written as `+10 <$> signal`

Author: Yeah and perhaps even more typically written as `signal + 10`, but I feel that distracts from the blog post.

  { name: "D",    wave: "22222",  data: ["A", "B", "C", "D", "E"] },
  { name: "Q",    wave: "22222",  data: ["init", "A", "B", "C", "D", "E"] },
]}
</script>

I.e., the output signal **Q** is a shifted version of **D**. In this diagram, the initial state of the register is marked **init**. A fourth primitive Clash offers is therefore `register`:

```haskell
-- | Simplified version of Clash's `register`: the actual version has more control signals.
register :: Clock dom -> a -> Signal dom a -> Signal dom a
register _clk init signal = init :- signal
```

Again, when Clash sees this primitive used, it generates a hardcoded bit of HDL. Note that because `register` operates on a single domain, it does not need to be aware of any properties of the clock: the abstraction works the same for clocks running on 1 MHz or 100 MHz. The generated HDL *does* need clock lines to be routed though, so the primitive still needs it.

And just like that, we've introduced state. We can now implement a simple counter:

```haskell
counter = register clock (fmap (+1) counter)

-- Because `Signal` has a `Num` instance, we can also write:
-- counter = register clock (counter + 1)
```

> From a HDL compilation point of view, we can see how this might work. If Clash encounters the special primitive `register`, it will not try to translate the definition, but it will insert a predefined bit of HDL.

Reviewer: ```suggestion
From an HDL compilation point of view, we can see how this might work. If Clash encounters the special primitive `register`, it will not try to translate the definition, but it will insert a predefined bit of HDL.
```

Unless of course you pronounce it _huddle_, which sounds like a fun pronunciation!

> From a HDL compilation point of view, we can see how this might work. If Clash encounters the special primitive `register`, it will not try to translate the definition, but it will insert a predefined bit of HDL.

Author: I'll obviously fix it, but _ter lering ende vermaak_: https://www.youtube.com/watch?v=nCe7Fj8-ZnQ. Seems like it might be going away!

> From a HDL compilation point of view, we can see how this might work. If Clash encounters the special primitive `register`, it will not try to translate the definition, but it will insert a predefined bit of HDL.

Reviewer: Oh that's a nice video, thanks!

  = 0 :-           let tmp = (0 + 1) :- tmp + 1 in tmp -- <=> eval (+)
  = 0 :-           let tmp = 1       :- tmp + 1 in tmp -- <=> inline tmp
  = 0 :- 1 :-      let tmp = (1 + 1) :- tmp + 1 in tmp -- <=> eval (+)
  = 0 :- 1 :-      let tmp = 2       :- tmp + 1 in tmp -- <=> inline tmp
  = 0 :- 1 :- 2 :- let tmp = (2 + 1) :- tmp + 1 in tmp -- <=> eval (+)
  = ..
```

### Signals on multiple domains
One of the questions `/u/netj_nsh` asked was: 

> Does Clash [..] support [..] synchronous design with clock domain crossing?

With the primitives introduced we cannot construct any clock domain crossing. In fact, the Haskell compiler will stop us before we even get a chance to run it:

```haskell
>>> a = pure 3 :: Signal A Int
>>> b = pure 5 :: Signal B Int
>>> liftA2 (+) a b
*** Type error: could not deduce A ~ B
```

And this is great! It means you can never accidentally cross clock domains in Clash. Of course, sometimes you *do* want to cross clock domains, which can be done with yet another primitive: `unsafeSynchronizer`:

```haskell
unsafeSynchronizer :: Clock dom1 -> Clock dom2 -> Signal dom1 a -> Signal dom2 a
```

> It will reinterpret a signal on domain `dom1` as if it were a signal on `dom2`. It keeps a measure called *relative time* that is keeps track of the distance to an edge. If relative time is less than zero, it represents the distance to the next edge of the consumed signal. If it is greater than zero, it represents the distance to the next edge of the produced signal. The former means an element should be produced. The latter means an element should be consumed. If relative time equals zero, both should happen. In Clash 1.0 this is programmed as:

Reviewer: ```suggestion
It will reinterpret a signal on domain `dom1` as if it were a signal on `dom2`. It keeps a measure called *relative time* that keeps track of the distance to an edge. If relative time is less than zero, it represents the distance to the next edge of the consumed signal. If it is greater than zero, it represents the distance to the next edge of the produced signal. The former means an element should be produced. The latter means an element should be consumed. If relative time equals zero, both should happen. While the current definition is more involved, Clash used to program this as:
```

The reference to 1.0 raised the question: why are we referring to a very old version when we're currently at 1.8? I think it's best to just deal with it and explain why what we say here actually doesn't correspond to what we do.

> It will reinterpret a signal on domain `dom1` as if it were a signal on `dom2`. It keeps a measure called *relative time* that is keeps track of the distance to an edge. If relative time is less than zero, it represents the distance to the next edge of the consumed signal. If it is greater than zero, it represents the distance to the next edge of the produced signal. The former means an element should be produced. The latter means an element should be consumed. If relative time equals zero, both should happen. In Clash 1.0 this is programmed as:

Author: > While the real definition in Clash is more involved, this computation boils down to this code:

I like that

> It will reinterpret a signal on domain `dom1` as if it were a signal on `dom2`. It keeps a measure called *relative time* that is keeps track of the distance to an edge. If relative time is less than zero, it represents the distance to the next edge of the consumed signal. If it is greater than zero, it represents the distance to the next edge of the produced signal. The former means an element should be produced. The latter means an element should be consumed. If relative time equals zero, both should happen. In Clash 1.0 this is programmed as:

Author: And I really like the updated description

> It will reinterpret a signal on domain `dom1` as if it were a signal on `dom2`. It keeps a measure called *relative time* that is keeps track of the distance to an edge. If relative time is less than zero, it represents the distance to the next edge of the consumed signal. If it is greater than zero, it represents the distance to the next edge of the produced signal. The former means an element should be produced. The latter means an element should be consumed. If relative time equals zero, both should happen. In Clash 1.0 this is programmed as:

Reviewer: In my new review I've included the suggestion for the rephrasing of the last sentence.

    {"name": "data (B)", "wave": "2x2", "data": [0, 2], "period": 1},
  ]
}
</script>

Though `Signal` can never represent this: it must have exactly one value representing a clock cycle.

### Signals: a leaky abstraction
In presence of multiple domains, signals become a leaky abstraction. For example, synchronizing a fast domain to a slower one loses information and resynchronizing to the fast domain does not necessarily give you back the same signal:

<script type="wavedrom">
{
  "signal": [
    {
      "name": "clock A",
      "wave": "l.hlhlhlhlhlhl",
      "period": 0.5
    },
    {
      "name": "clock B",
      "wave": "l.h..l..h..l..",
      "period": 0.5
    },
    {
      "name": "A",
      "wave": "2222222",
      "data": [
        0,
        1,
        2,
        3,
        4,
        5,
        6
      ],
      "period": 1
    },
    {
      "name": "A => B",
      "wave": "22..2..",
      "data": [
        0,
        3,
        6
      ],
      "period": 1
    },
    {
      "name": "A => B => A",
      "wave": "2222222",
      "data": [
        0,
        3,
        3,
        3,
        6,
        6,
        6,
      ],
      "period": 1
    }
  ]
}
</script>

> In the experience of your's truly, this is not a problem in practise: synchronization with `unsafeSynchronizer` is almost always backed by a memory element, sidestepping the problem entirely.

Reviewer 3: ```suggestion
In the experience of yours truly, this is not a problem in practise: synchronization with `unsafeSynchronizer` is almost always backed by a memory element, sidestepping the problem entirely.
```

Reviewer: ```suggestion
In the experience of your's truly, this is not a problem in practice: synchronization with `unsafeSynchronizer` is almost always backed by a memory element, sidestepping the problem entirely.
```

Author: I always have to look up practise/practice :(

## Recap
To recap, `Signal` works for HDL translation because its implementation is hidden from the public API. To make the data type useful still, primitives are offered:

* `pure`: construct a signal
* `<$>` or `fmapSignal`: apply a pure function/combinational circuit to a signal
* `<*>` or `appSignal`: merge multiple signals on the same domain
* `register`: introduce state

For simulation it works because:

 * Lazy evaluation means we can create feedback loops without compilation or runtime errors.
 * We can "synchronize" signals on multiple domains by calculating their relative clock offsets and constructing a new signal.

## Footnotes
<sup>1</sup>This isn't entirely true: Haskell provides "escape hatches" such as `unsafePerformIO` that allow you to break purity.

<!-- <sup>2</sup>In theory Clash could analyze `map`, determine how often it uses its function argument, and render that as input and outputs. At the time of writing, Clash does not do this. -->

> <sup>2</sup>These three operations effectively implement `Applicative` (and `Functor`).

Reviewer 2: ```suggestion
<sup>2</sup>These three operations effectively implement the [Applicative](https://hackage.haskell.org/package/base-4.21.0.0/docs/Prelude.html#t:Applicative) and [Functor](https://hackage.haskell.org/package/base-4.21.0.0/docs/Prelude.html#t:Functor) typeclasses.
```
